core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -230,6 +227,7 @@ private void signalAsyncReadComplete() {
 
   private void waitForAsyncReadComplete() throws IOException {
     stateChangeLock.lock();
+    isWaiting.set(true);
     try {
       while (readInProgress) {
```

Comment:
shall we add a comment about spurious wakeup? Otherwise someone else may still mistakenly remove it in the future.